### DISCUSSION TITLE
Does Milvus load every vector in a collection into memory on search?

### DISCUSSION THREAD

**Discussion by shoumikchow** [milvus-io/milvus#16484] [Q&A and General discussion] [answered]
Hi,

I have a question regarding memory usage for the `load()` operation performed prior to a search.

Milvus' docs say that the entire collection is loaded into memory (https://milvus.io/docs/v2.0.x/search.md#Load-collection) before performing a search. Does this mean that every vector in the collection gets loaded into memory, e.g. RAM?

If every vector in a collection gets loaded, do I need to make sure that the machine hosting Milvus has enough RAM to load the entire collection, or is there some optimization happening under the hood?

Thanks!

**Comment by yhmo** [comment, marked as answer]
Yes, it requires RAM can hold entire collection. The behavior of load_collection():
1. for segment that has no index, load every vector of the segment into RAM
2. for segment that has index, load index data into RAM

We call vectors as "raw data". Usually index data size is not equal to raw data. For example, IVFLAT is a bit larger than raw data, IVFPQ is about 25% of raw data.

After you call create_index(), milvus build index in background thread. Then you call load_collection, milvus load segments(raw data or index data) into RAM. Once a segment finish index, milvus will automatically load index data to replace the raw data(free raw data from memory).

**Reply by dsp05** [reply]
Hi, what happens to the on disk index (DiskANN)?

**Reply by xiaofan-luan** [reply]
> Hi, what happens to the on disk index (DiskANN)?

Disk index only load quantitive index and original data resident on NVMe disk

**Reply by dsp05** [reply]
> > Hi, what happens to the on disk index (DiskANN)?
> 
> Disk index only load quantitive index and original data resident on NVMe disk

Thanks for the reply.! A followup question, what is size of the "quantitive index" compared to the raw data? If IVFPQ is 25%, I assume DiskANN would use much less than that?

**Reply by xiaofan-luan** [reply]
> > > Hi, what happens to the on disk index (DiskANN)?
> > 
> > 
> > Disk index only load quantitive index and original data resident on NVMe disk
> 
> Thanks for the reply.! A followup question, what is size of the "quantitive index" compared to the raw data? If IVFPQ is 25%, I assume DiskANN would use much less than that?

Usually is among 1:5 or 1:6, but you can tune the parameters. The less memory usage, the higher iops required for same Recall.

**Reply by RakeshRaj97** [reply]
How to tune the parameters? Can you provide a code snippet for example?

**Reply by yhmo** [reply]
@RakeshRaj97 
Read the doc about DiskANN first: https://milvus.io/docs/disk_index.md#DiskANN-related-Milvus-configurations

DiskANN also requires memory, the in-memory part actually is PQ index, the on-disk part is a [Vamana graph](https://zilliz.com/learn/DiskANN-and-the-Vamana-Algorithm).

So, from the doc you can these parameters:
- MaxDegree
- SearchListSize:
- PQCodeBugetGBRatio
- SearchCacheBudgetGBRatio
- BeamWidthRatio

All these parameters are configured in the milvus.yaml: https://github.com/milvus-io/milvus/blob/3b5ce73ded1003e881a0c39ccc61f223bc070106/configs/milvus.yaml#L563
**Modify the milvus.yaml and restart milvus so that it takes effect. These parameters cannot be changed by client code.**
Read this doc to know how to modify the milvus.yaml and take effect: https://milvus.io/docs/configure-docker.md

From the description, we know:
- Increase MaxDegree value, the recall rate increase, the size of Vamana graph increase(which means the on-disk part increase), and build index time increase
- Increase SearchListSize value, the recall rate increase, build index time increase
- Increase PQCodeBugetGBRatio value, the recall rate increase, the size of in-memory PQ index also increase
- Increase SearchCacheBudgetGBRatio value, memory usage of search increase, build index time decrease

So, if you want to save memory, decrease the value of PQCodeBugetGBRatio and SearchCacheBudgetGBRatio.
If you want to save disk space, decrease the value of MaxDegree.
But note that decreasing these values will get poor recall rate for your business.

**Reply by RakeshRaj97** [reply]
Thanks for the explanation. I adjusted the above values for DiskANN and see a high memory usage in querynodes while loading the collection. I have a collection of 100million vectors of 2048 dimensions. I am using DiskANN as index method. after inserting and building the index, now I am trying to load the collection for querying. My node has memory capacity of 1TB and 64 CPUs. My understanding with DiskANN approach was to use disk instead of memory for large datasets. Please suggest some ideas.

**Comment by cancerberoSgx** [comment]
So does this means there is a hardware limit for the amount of embeddings I can (similarity) search ? 
Here I'm trying to figure out if milvus would support 500k+ embeddings/vector similarity search or even more. 
Is it possible to handle this problem? or is just something milvus was not designed to support?
thanks

**Reply by yhmo** [reply]
500k is a pretty small dataset.
Milvus is in-memory search database, just make sure the RAM resource is sufficient for 500k embeddings.

Let's say your embeddings are 1024 dimensional, the original data size for 500k embeddings is 500k*1024*4bytes=2GB.
If you chose IVF_FLAT/HNSW index, the required RAM size for query nodes is about 2GB+
If you chose IVF_PQ/IVF_SQ8, the required RAM size for query node is about 600MB+

**Reply by cancerberoSgx** [reply]
thanks @yhmo for the answer. I was having a memory error, but because I was trying to .insert() 500k entries in a single call and I thought it was because of collection load(). Inserting entries in chunks solved the problem. Thanks again! BTW: I'm comparing similarity search of milvus vs chromadb, and for these large catalogs, milvus seems to be doing much much better. :)

**Reply by rere950303** [reply]
@yhmo
> If you chose IVF_FLAT/HNSW index, the required RAM size for query nodes is about 2GB+

If the querynode replica is 5, you mean i need 2GB of total memory by combining 5 querynodes, right?
